[HDM-28] searchbar component refactor / hover/active state fix

adrian-oleskiewicz opened the pull request:
## Issue tracking
<!--
  Where is the progress of this issue being tracked?
  Where can one find the requirements for this issue?
  Where did this issue originated from?

  E.g. GitHub issue, Slack message, etc.
-->
HDM-28
[UPDATE] HDM-31 - implemented redesign of header
## Context behind the change
<!--
  What changes have you made to the code, and why?

  Describe the goal of this pull request. What does it change or fix?
  At a high level, what parts of the code did you change and why?
-->
I had to refactor the code to stick to the DRY principle. I created 1 shareable component from 2 components. I moved some mui styles to separate file to improve readability
## How has this been tested?
<!--
  You should always test your changes.

  Describe the steps you took to make sure your PR does what it's supposed to do.
  How we ensure that adjacent code/features are still working?
  Are there any performance implications?
-->
I tested all searchbars where I made changes to the code and compared the local UX with the current one at https://dashboard.humanprotocol.org/
## Release plan
<!--
  If not you, but somebody else will be deploying this, what they need to know/do to succeed?

  Add any action items that need to be done for the release (any step, before/during/after),
  e.g. running a DB migration, leaving a note about release in Slack, adding new envs, etc.
-->
normal deploy
## Potential risks; What to monitor; Rollback plan
<!--
  What might go wrong with deployment of this PR?
  Can it affect some other services/areas? In what way?
  What metrics/dashboards/etc. you should keep an eye on while/after deploying this?
 
  How will we handle any issues if they arise? Do we need rollback plan?
  Do we need a second pair of eyes on this?
-->

dnechay reviewed (commented):
Nice refactoring

Some pieces have missed: when you select some network and hit "search" - chainId in URL is not updated, and from what I see it seem like `137` is always queried.

Also on PRD search bars are in sync, i.e. when you change value on page - it's also changed on top bar, which is not the case in this PR. Please check it

dnechay commented on packages/apps/dashboard/ui-2024/src/components/SearchBar/SearchBar.tsx:
```diff
@@ -0,0 +1,199 @@
+
+  const handleSelectChange = (event: SelectChangeEvent<string>) => {
+    const chainId = Number(event.target.value);
+    const networkName = getNetwork(chainId)?.name || '';
```
As function itself - yes, it can. But for `Select` component we provide a specific list of `networks` that is already filtered by `getNetwork` (to have chain name and id). So IMU `getNetwork` will return `undefined` only if `chainId` selected is empty, so I though maybe we could improve something there as well. But taking into account that we have different `renderValue` for empty state - it doesn't makes sense to touch it

dnechay reviewed (commented):
One item to address - and should be good to go.

When you have some `address` entered and change networks (e.g. from "Polygon" to "Ethereum") page updates with results for selected network (and it's in sync with the top search bar), but `chainId` in page url is still the same, so when you reload page you get to not last selected chain, but to one in the url. So probably we need to run `navigate` also when selecting valid network

dnechay reviewed (commented):
Also please pay attention that it should have a bit different theming on mobile devices

Expected (from Figma)
<img width="828" alt="Screenshot 2024-12-03 at 14 17 11" src="https://github.com/user-attachments/assets/5e2f7e3e-55b0-4b54-9222-f8813f80f24c">

Actual
<img width="338" alt="Screenshot 2024-12-03 at 14 17 59" src="https://github.com/user-attachments/assets/8fa3396f-e87b-493c-a51e-390f0d7ddf0d">

adrian-oleskiewicz commented:
> One item to address - and should be good to go.
> 
> When you have some `address` entered and change networks (e.g. from "Polygon" to "Ethereum") page updates with results for selected network (and it's in sync with the top search bar), but `chainId` in page url is still the same, so when you reload page you get to not last selected chain, but to one in the url. So probably we need to run `navigate` also when selecting valid network

@dnechay I fixed this one. Additionally I fixed jumping content on loading state :)

dnechay commented:
@adrian-oleskiewicz

>  but chainId in page url is still the same

It's fixed 👍🏻 

> design adjusted

it still has purple border even when not in focus state

Also when you on the main page, have empty input and change the network - it moves you on "search results" page with `0x0`, but:
- it should do it only if there is value
- it should keep it empty
<img width="305" alt="Screenshot 2024-12-03 at 16 35 15" src="https://github.com/user-attachments/assets/edaf8b73-f512-4a8b-916b-a2e20e135c24">

Maybe we apply search only in case if user hits enter/search icon?

The expected behavior would be next:
- search bars are in sync
- user should never see `0x0`, empty value if none (we have a component for it)
- user can input address and select network nothing happens
- search is run only when input is submitted

wdyt?

adrian-oleskiewicz commented:
> @adrian-oleskiewicz
> 
> > but chainId in page url is still the same
> 
> It's fixed 👍🏻
> 
> > design adjusted
> 
> it still has purple border even when not in focus state
> 
> Also when you on the main page, have empty input and change the network - it moves you on "search results" page with `0x0`, but:
> 
> * it should do it only if there is value
> * it should keep it empty
> 
> <img alt="Screenshot 2024-12-03 at 16 35 15" width="305" src="https://private-user-images.githubusercontent.com/15381780/391985531-edaf8b73-f512-4a8b-916b-a2e20e135c24.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMyMzM1NTQsIm5iZiI6MTczMzIzMzI1NCwicGF0aCI6Ii8xNTM4MTc4MC8zOTE5ODU1MzEtZWRhZjhiNzMtZjUxMi00YThiLTkxNmItYTJlMjBlMTM1YzI0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjAzVDEzNDA1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU0YjIyNjk5NDcyNDNkNmJlYzQ2YWVkMTI2NWY5ZDExZmE4YjBkMDFmMWMwYzIxNWI1YjM4MDM2ZGNhNDg2ZTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Xf1Q2WUmtqNTeQZLzRMps9Pp15iSy1_p2_oWsle694w">
> Maybe we apply search only in case if user hits enter/search icon?
> 
> The expected behavior would be next:
> 
> * search bars are in sync
> * user should never see `0x0`, empty value if none (we have a component for it)
> * user can input address and select network nothing happens
> * search is run only when input is submitted
> 
> wdyt?

Ok now I fully understand, thanks! :) 
1. I improved this input border and there is active state now
2. When user click on input X button to clear address than we redirect to main page
3. If the user enters an address, then searches, then removes the address with a backspace, and tries to change the chain without entering a new address, then I paste the address deleted with the backspace.
4.If the user does not enter the address and does not have an address in the URL, and changes the chain, we do not send the query

dnechay commented:
@adrian-oleskiewicz
> I improved this input border and there is active state now

👍🏻 

> When user click on input X button to clear address than we redirect to main page

TBH, that's not intuitive and feels buggy. As a user I would expect that when you hit `X` - input is cleared, but nothing else happens. i.e. no redirect, no new query. The reason is next: as a user a just want to clear the input to insert/type new value, but if I get redirected -> it means loading data from new page (which might be annoying in slow internet) and makes me look for the "Search bar" on page again. So let's just erase input when `X` is clicked

> If the user enters an address, then searches, then removes the address with a backspace, and tries to change the chain without entering a new address, then I paste the address deleted with the backspace

That's not necessary, please let's keep it empty. Reasoning is similar to the above: I clean the input, then want to change network and paste new value, but if we put the old one - it makes me clear it again

> If the user does not enter the address and does not have an address in the URL, and changes the chain, we do not send the query

That's nice!

adrian-oleskiewicz commented:
> @adrian-oleskiewicz
> 
> > I improved this input border and there is active state now
> 
> 👍🏻
> 
> > When user click on input X button to clear address than we redirect to main page
> 
> TBH, that's not intuitive and feels buggy. As a user I would expect that when you hit `X` - input is cleared, but nothing else happens. i.e. no redirect, no new query. The reason is next: as a user a just want to clear the input to insert/type new value, but if I get redirected -> it means loading data from new page (which might be annoying in slow internet) and makes me look for the "Search bar" on page again. So let's just erase input when `X` is clicked
> 
> > If the user enters an address, then searches, then removes the address with a backspace, and tries to change the chain without entering a new address, then I paste the address deleted with the backspace
> 
> That's not necessary, please let's keep it empty. Reasoning is similar to the above: I clean the input, then want to change network and paste new value, but if we put the old one - it makes me clear it again
> 
> > If the user does not enter the address and does not have an address in the URL, and changes the chain, we do not send the query
> 
> That's nice!

@dnechay Ok, I see :) I rolled back this behavior. 
I suggested that when you click the X button, the deleted address remains in the URL and if you refresh, the UI state is different

dnechay commented:
@adrian-oleskiewicz 

> I suggested that when you click the X button, the deleted address remains in the URL and if you refresh, the UI state is different

That's fine, because at that moment user is on "results" page (specific one), so we should stay there, but still clean input

dnechay commented:
@adrian-oleskiewicz 

> I suggested that when you click the X button, the deleted address remains in the URL and if you refresh, the UI state is different

But that means only cleaning an input and still showing previous result, not `Nothing found :(`, so:
- user makes some search
- they are on "results" page
- they click on `X`
- input is cleared, but user still sees old result and previous address is in url (so results and url are in sync)

adrian-oleskiewicz commented:
> @adrian-oleskiewicz
> 
> > I suggested that when you click the X button, the deleted address remains in the URL and if you refresh, the UI state is different
> 
> But that means only cleaning an input and still showing previous result, not `Nothing found :(`, so:
> 
> * user makes some search
> * they are on "results" page
> * they click on `X`
> * input is cleared, but user still sees old result and previous address is in url (so results view and url are in sync)

@dnechay Ok it makes sense, thanks :) Do you see any other improvements for this input ?

dnechay commented:
@adrian-oleskiewicz

> Do you see any other improvements for this input ?

In terms of behavior or current PR - nope. I will need to check the code when behavior is adjusted, but expect it to be fine

dnechay reviewed (commented):
In general - LGTM

Let's change the behavior to what mentioned/agreed already - and it should be good to go

adrian-oleskiewicz commented:
@dnechay I removed this behavior and now when user clicks clear button (X) it want fetch data
